### DISCUSSION TITLE
Windows and Linux different results. No color information with composite?

### DISCUSSION THREAD

**Discussion by audas** [ImageMagick/ImageMagick#4517] [Help]
I am using VERSION 7.1 on both windows and Linux - and layering images preserving alpha channels. When I do this on windows its fine, but on Linux the result is black and white?

If the first layer has alpha set above zero (has some alpha) then it works, but at zero it is always black and white?

convert  -channel a -evaluate multiply 1 +channel   ../images/made_images/61a4323b5d68e/undo/preview_61a4323b5d68e_.png ../images/made_images/61a4323b5d68e/undo/layer_61a43245c7784_.png  -size 1024x512 -gravity center   -composite ../images/made_images/61a4323b5d68e/undo/preview_61a4323b5d68e_.png 

Anyone with some insight into this?

**Comment by fmw42** [comment]
Syntax is not proper. Read the input first.  Syntax order is even more important in IM 7.

**Reply by audas** [reply]
Thanks Fred - I am converting that first input - if we change the order then there is no alpha applied.

Also it should not work on windows? Surely it would be the same?

So if we move the command "-channel a -evaluate multiply 1 +channel" to after the first input then it does not apply its settings.

Edit:

Looks like others have encountered similar issues. 

https://legacy.imagemagick.org/discourse-server/viewtopic.php?t=33400

**Reply by fmw42** [reply]
What exactly are you trying to do? Please explain .  Can you show the input images and what you expect for the output.

**Comment by audas** [comment]
![image](https://user-images.githubusercontent.com/274994/143802289-aafe53bd-10d0-42ff-9937-0b2feb137f8b.png)

![image](https://user-images.githubusercontent.com/274994/143802321-7ea4261f-b8b4-446e-be2d-89c45e5b115b.png)

Here they are blended with the lower layer at 31% 

![image](https://user-images.githubusercontent.com/274994/143802404-0bb5aae4-8aed-47e1-b091-905b322f6372.png)


Then as soon as I set the lower layer to ZERO it removes all colour information

![image](https://user-images.githubusercontent.com/274994/143802485-db0f47ec-6590-43a7-ba39-44409aba68ee.png)

**Comment by audas** [comment]
I am adding layers with varying levels of "alpha" - so that as each layer is added it preserves its transparency.

I first create a basic background with color and transparency, then I am looping through "layers" and outputting them, then adding the next via PHP.

So the last image above should just be the coloured balloons with the red squiggle line completely invisible.

Thanks for the time.

**Reply by fmw42** [reply]
Sorry, I still do not understand exactly what you want, step by step. Suppose we call the images balloon.png and squiggle.png.  Tell me exactly what you want to happen at each step of the layering, one step at a time.  Also why do you want the squiggle to be invisible? What is the point of that? I note that -size 1024x512 does nothing as there is no constant xc: or canvas following it.

**Reply by fmw42** [reply]
I also note that -channel a -evaluate multiply 1 +channel, when operating on an image will do nothing. Multiplying by 1 is a no-op action. It keeps the channel the same.

**Reply by fmw42** [reply]
You appear to be trying to use Unix syntax. But ImageMagick syntax is not the same as Unix syntax. See https://legacy.imagemagick.org/Usage/basics/#why.

**Reply by audas** [reply]
I am outputting multiple layers (there could be many) each one is added to the previous with its transparency.

Thats it. 

Your questions:
"Why show an invisible layer"
This is coming from an interface, so users may have set that layer to 0 for some reason.

"-size 1024x512" does nothing

This is setting the image to the size required. Works.


"channel a -evaluate multiply 1 +channel, when operating on an image will do nothing."

This is also definitely working as you can see by the intermediate image which is set at 31% and the lower layer is "faded"

"You appear to be trying to use Unix syntax. "

Also everything is working when things are set to any level ABOVE 0% transparency. Has always worked on previous versions of imagemagick and is only now having issues.

Fundamentally this is an issue with the lower layer set at ZERO.

**Reply by fmw42** [reply]
Your syntax is not correct!  -size does nothing as you have images that have their own size. If you want to change and image's size use -resize not -size.

I cannot say what you are doing in PHP Imagick. But in command line, your syntax is not OK.

**Comment by fmw42** [comment]
Suppose I composite the squiggle over the balloon using different alpha settings. Here is a composite with alpha multiplied by 1, 0.5 and 0. In Unix syntax, this would be

`convert balloon.png \( squiggle.png -channel a -evaluate multiply 1 +channel \) -compose over -composite result1.png `


![result1](https://user-images.githubusercontent.com/13458218/143805328-363fd568-5ec8-41c2-abbe-2ac3298f3723.png)

`convert balloon.png \( squiggle.png -channel a -evaluate multiply 0.5 +channel \) -compose over -composite result2.png `


![result2](https://user-images.githubusercontent.com/13458218/143805335-2ed54402-363c-4257-a260-02ea2abd4a32.png)

`convert balloon.png \( squiggle.png -channel a -evaluate multiply 0 +channel \) -compose over -composite result3.png `

![result3](https://user-images.githubusercontent.com/13458218/143805340-9cfa2ebb-59b3-414d-84e7-d314d056e62d.png)


Note that I use parenthesis processing to avoid the -evaluate from processing on the balloon image.  

See https://legacy.imagemagick.org/Usage/basics/#parenthesis

In Windows, the command would be the same, except do not escape the parentheses with back-slash.

Does this resolve the issue?

**Comment by snibgo** [comment]
As fmw42 says, although I think the OP wants the squiggle under the balloons, not on top.

In the original command, try inserting "-set colorspace sRGB" after reading the second image, before the "-composite".

This is because I suspect your problem is that the first image "as soon as I set the lower layer to ZERO" becomes grayscale, and that metadata is passed to the output of "-composite", so "it removes all colour information".

**Reply by fmw42** [reply]
If you want to composite it under use +swap. If for some reason it becomes grayscale (when the first image is fully transparent -- the squiggle, after swapping), then as snibgo says, force the image to be color by -set colorspace sRGB

For me, swapping (to slide the squiggle under the balloons) still provides a color output:

`convert balloon.png \( squiggle.png -channel a -evaluate multiply 0 +channel \) +swap -compose over -composite result4.png`

![result4](https://user-images.githubusercontent.com/13458218/143806186-51b56553-9668-452f-9bf0-fed9cba5200b.png)

**Reply by audas** [reply]
Ok - so snibgo's advice has brought the colour back however it is now placing the color ABOVE the layer ABOVE it. 

![image](https://user-images.githubusercontent.com/274994/143806346-66230bfa-e7a7-4170-a59f-75fd3838e458.png)


Ignore green cross, added that for testing 3 layers.

**Comment by audas** [comment]
As an update I have been unable to perform this function with any combination I have seen this is probably due to there being more than two or three layers - I have tested all combinations suggested here.

The only thing which FIXED the black and white issue was the "-set colorspace sRGB " suggestion.

My solution was to apply the transparency to the layer into a temporary file - and then use that temp file to -compose overlay.

If anyone can resolve this issue for WINDOWS and LINUX that would be great.

**Reply by fmw42** [reply]
Post an example command line with 3 images to layer. Did my solution not work for layering 2 images? You have to do that twice with 3 images. -composite works in pairs of images (not more than 2).

**Comment by snibgo** [comment]
> Ok - so snibgo's advice has brought the colour back however it is now placing the color ABOVE the layer ABOVE it.

I'm glad something works. I don't understand "placing the color ABOVE the layer ABOVE it." You don't show your command, so it is difficult to say what is wrong. And I don't really understand what you want, which makes it more difficult.

Perhaps what you want is this: you have a number of layers. You want to adjust the alpha of any layer between 0 (fully transparent)and 1 (fully opaque), then composite them over each other.

Windows BAT script:
```
magick ^
  ( squiggle.png -channel A -evaluate Multiply 0.31 +channel -set colorspace sRGB ) ^
  ( toes.png -alpha set -channel A -evaluate Multiply 0.6 +channel -set colorspace sRGB ) ^
  ( balloons.png -channel A -evaluate Multiply 1 +channel -set colorspace sRGB ) ^
  -background None ^
  -compose Over -layers merge ^
  out.png
```
The file toes.png has no alpha channel, so I need to set it before I can multiply the alpha channel. I could put "-alpha set" after any input png.

**Reply by audas** [reply]
Just for clarity - the number of layers in my use case is unknown - potentially 10 or even more.

So I must loop through the information for each layer and add that layer to an output image. I then add the next layer to the existing image and so forth until (....N) number of images/layers have been added.

Looking at snibgo's last post - this may well work as the layers were "losing" their alpha levels etc.

Here is something which provides a basic idea.

```
foreach ($_POST['layers'] as $layer) {
        $new_layer=$layer;

convert   $new_layer_.png  -channel a -evaluate multiply 1 +channel     -size 1024x512  temp_layer.png

convert   final_.png temp_layer.png  -size 1024x512 -gravity center -set colorspace sRGB    -compose exclusion -composite final_.png

}
```
This produces as many layers as needed, each layer set at its own transparency and then merged. Notice the "-copose exclusion" - each layer can also have its own blend mode as well.

Each layer can also have its own "transformation" applied to it as well (rotate, scale, skew, flip, mirror etc) - so if anyone is interested here is a complete method for doing this to unlimited layers in PHP.

This allows a user to take the transformation applied to an image in a browser window from the CSS transformations, CSS or ImageMagick "blend modes" and any alpha settings and add each layer with its own transformation etc onto a final image and output it.


For completeness and anyone else seeing this in the future.

```
$_POST['layers'][0]['src']=strtok($_POST['layers'][0]['src'],"?" );
    $info=pathinfo($_POST['layers'][0]['src']);
    $dims = getDimensions($_POST['layers'][0]['src']);
    $dir=$info['dirname'];
    if (!is_dir($dir))mkdir($dir, 0755, true);
    $dir=str_replace("/undo", "", $dir);
        if (!is_dir($dir."/undo/"))mkdir($dir."/undo/", 0755, true);
    $output=$dir."/undo/preview_".$_POST['uniqid']."_.".$info['extension'];

 $bg=$dir."/undo/bg_".$_POST['uniqid']."_.".$info['extension'];
    $img = @imagecreatetruecolor ( $dims['image_width'], $dims['image_height'] );
    $black = imagecolorallocate($img, 0, 0, 0);
    imagecolortransparent($img, $black);
    imagepng ( $img, $bg );

    foreach ($_POST['layers'] as $layer) {
        $layer['src']=strtok($layer['src'],"?" );
        $compose="";
        $alpha="";
        if(isset($layer['alpha']) && $layer['alpha']!="" && $layer['alpha']!="none"){
            $alpha=" -channel a -evaluate multiply ".$layer['alpha']." +channel  ";
            
        }
        if(isset($layer['blendmode']) && $layer['blendmode']!="" && $layer['blendmode']!="none")$compose=" -compose ".$layer['blendmode'];
        if(isset($layer['transform']) && $layer['transform']!=""  && $layer['transform']!="none" && $layer['transform_full']!="none"){
            
            $layer['transform_full']=str_replace("matrix(", "", $layer['transform_full']);
            $layer['transform_full']=str_replace(")", "", $layer['transform_full']);
            $layer['transform_full']=str_replace(" ", "", $layer['transform_full']);
            $layer['transform']['scale'][0]=str_replace(" ", "", $layer['transform']['scale'][0]);
            $layer['transform']['scale'][1]=str_replace(" ", "", $layer['transform']['scale'][1]);
            $layer['transform']['translate'][0]=str_replace(" ", "", $layer['transform']['translate'][0]);
            $layer['transform']['translate'][1]=str_replace(" ", "", $layer['transform']['translate'][1]);
            
            $geometry=$dir."/undo/bg_1_".$_POST['uniqid']."_.".$info['extension'];
            $img = @imagecreatetruecolor ( $dims['image_width'], $dims['image_height'] );
            $black = imagecolorallocate($img, 0, 0, 0);
            imagecolortransparent($img, $black);
            imagepng ( $img, $geometry );
            
            
            $cmd="convert ".$layer['src']." -matte -virtual-pixel Transparent  -affine  \"".$layer['transform_full']."\" -transform ".$geometry."  2>&1";
            
            $result=exec($cmd);
            $geometry_2=$dir."/undo/bg_2_".$_POST['uniqid']."_.".$info['extension'];
            $img = @imagecreatetruecolor ( $dims['image_width'], $dims['image_height'] );
            $black = imagecolorallocate($img, 0, 0, 0);
            imagecolortransparent($img, $black);
            imagepng ( $img, $geometry_2 );
            $final_geometry=$dir."/undo/bg_3_".$_POST['uniqid']."_.".$info['extension'];
            
            $dim_1 = getDimensions($geometry);
            $wd=$layer['transform']['translate'][0]+($dim_1["image_width"]*.5);
            $ht=$layer['transform']['translate'][1]+($dim_1["image_height"]*.5);
            $wd=$layer['pos'][0];
            $ht=$layer['pos'][1];
            $cmd="convert  ".$geometry_2." ".$geometry." -geometry +".$wd."+".$ht." -composite  ".$final_geometry." 2>&1";
            $result=exec($cmd);
            $layer['src']=$final_geometry;
        }
        

        $temp_layer_info=pathinfo($layer['src']);
        $temp_layer=$temp_layer_info['dirname'].uniqid().".".$temp_layer_info['extension'];
        $cmd="convert   ".$layer['src']." ".$alpha."   -size ".$dims['image_width']."x".$dims['image_height']."  ".$temp_layer." 2>&1";
        
        $result=exec($cmd);
        $cmd="convert   ".$bg." ".$temp_layer."  -size ".$dims['image_width']."x".$dims['image_height']." -gravity center -set colorspace sRGB   ".$compose." -composite ".$output." 2>&1";
        
        $result=exec($cmd);
        $bg=$output;
        unlink($temp_layer);
    }
```
